webapp/src/main/java/com/box/l10n/mojito/service/tm/TMTextUnitStatisticService.java

```diff
@@ -0,0 +1,154 @@
+import com.box.l10n.mojito.service.tm.search.StatusFilter;
+import com.box.l10n.mojito.service.tm.search.TextUnitDTO;
+import com.box.l10n.mojito.service.tm.textunitdtocache.TextUnitDTOsCacheService;
+import com.box.l10n.mojito.service.tm.textunitdtocache.UpdateType;
+import com.google.common.collect.ImmutableMap;
+import io.micrometer.core.instrument.MeterRegistry;
+import io.micrometer.core.instrument.Tags;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+import static java.util.stream.Collectors.groupingBy;
+
+/**
+ * @author garion
+ */
+@Service
+public class TMTextUnitStatisticService {
+    /**
+     * logger
+     */
+    static Logger logger = LoggerFactory.getLogger(TMTextUnitStatisticService.class);
+
+    @Autowired
+    MeterRegistry meterRegistry;
+
+    @Autowired
+    TextUnitBatchMatcher textUnitBatchMatcher;
+
+    @Autowired
+    TextUnitDTOsCacheService textUnitDTOsCacheService;
+
+    @Autowired
+    TMTextUnitRepository tmTextUnitRepository;
+
+    @Autowired
+    TMTextUnitStatisticRepository tmTextUnitStatisticRepository;
+
+    @Autowired
+    TextUnitUtils textUnitUtils;
+
+    /**
+     * Imports statistics information for a target text unit by matching it using name, content and comment information.
+     * Existing statistic data is overwritten by incoming data.
+     *
+     * @param locale                the locale of the target text unit
+     * @param asset                 the asset of the target text unit
+     * @param textUnitStatistics    the statistics of the text unit
+     * @return                      a PollableFuture
+     */
+    @Pollable(async = true, message = "Start importing text unit statistics.")
+    public PollableFuture importStatistics(
+            Locale locale,
+            Asset asset,
+            List<ImportTextUnitStatisticsBody> textUnitStatistics) {
+        logger.debug("Import statistics for the text units for the given locale and asset.");
+
+        ImmutableMap<String, TextUnitDTO> textUnitDTOsForAssetAndLocaleByMD5 = textUnitDTOsCacheService.getTextUnitDTOsForAssetAndLocaleByMD5(
+                asset.getId(),
+                locale.getId(),
+                StatusFilter.ALL,
+                true,
+                UpdateType.ALWAYS);
+
+        // Map ImportTextUnitStatisticsBody data to TextUnitDTO objects by their MD5 of (name + content + comment)
+        Map<String, ImportTextUnitStatisticsBody> md5ToTextUnitStatisticMap = new HashMap<>();
+        List<Long> textUnitIds = new ArrayList<>();
+        textUnitStatistics.forEach(textUnitStatistic -> {
+            String statisticTextUnitMD5 = getTextUnitMd5(textUnitStatistic);
```

Comment:
in some case we won't have the full information to compute the md5 (at runtime no having the source string). Should this support partial matching?

Reply:
Added fallback matching by name and content when missing comment info.